lib/useOnyx.ts

```diff
@@ -0,0 +1,188 @@
+                return;
+            }
+        } catch (e) {
+            throw new Error(
+                `'${previousKey}' key can't be changed to '${key}'. useOnyx() only supports dynamic keys if they are both collection member keys from the same collection e.g. from 'collection_id1' to 'collection_id2'.`,
+            );
+        }
+
+        throw new Error(
+            `'${previousKey}' key can't be changed to '${key}'. useOnyx() only supports dynamic keys if they are both collection member keys from the same collection e.g. from 'collection_id1' to 'collection_id2'.`,
+        );
+    }, [previousKey, key]);
+
+    const getSnapshot = useCallback(() => {
+        // We get the data from the cache, supplying a selector too in case it's defined.
+        // If `newData` is `undefined` it means that the cache doesn't have a value for that key yet.
+        // If `newData` is `null` or any other value if means that the cache does have a value for that key.
+        // This difference between `undefined` and other values is crucial and it's used to address the following
+        // conditions and use cases.
+        let newData = getCachedValue(key, selectorRef.current);
+
+        // Since the fetch status can be different given the use cases below, we define the variable right away.
+        let newFetchStatus: FetchStatus | undefined;
+
+        // If we have pending merge operations for the key during the first connection, we set data to `undefined`
+        // and fetch status to `loading` to simulate that it is still being loaded until we have the most updated data.
+        // If `allowStaleData` is `true` this logic will be ignored and cached data will be used, even if it's stale data.
+        if (isFirstConnectionRef.current && Onyx.hasPendingMergeForKey(key) && !options?.allowStaleData) {
+            newData = undefined;
+            newFetchStatus = 'loading';
+        }
+
+        // If data is not present in cache (if it's `undefined`) and `initialValue` is set during the first connection,
+        // we set data to `initialValue` and fetch status to `loaded` since we already have some data to return to the consumer.
+        if (isFirstConnectionRef.current && newData === undefined && options?.initialValue !== undefined) {
+            newData = options?.initialValue as CachedValue<TKey, TReturnData>;
+            newFetchStatus = 'loaded';
+        }
+
+        // If the previously cached data is different from the new data, we update both cached data
+        // and the result data to be returned by the hook.
+        // We can't directly compare the value from `dataRef` with `newData` because we default `undefined`
+        // to `null` when setting `dataRef` value to ensure we have a consistent return, but we need to be able to differentiate
+        // between `undefined` and `null` during the comparison, so `cachedValueRef` is used to store the real value without changing
+        // to `null`.
```

Comment:
This part seems quite complicated, had to read through this couple of times before I understood what is going on 😅 

One question, why can't we use null value for the loading status as well? This way we could simplify the logic a bit here

Reply:
Let's just simplify then, wdyt?

```ts
        // If the previously cached data is different from the new data, we update both cached data
        // and the result data to be returned by the hook.
        if (!deepEqual(cachedValueRef.current, newData)) {
            cachedValueRef.current = newData as CachedValue<TKey, TReturnData>;

            // If the new data is `undefined` we default it to `null` to ensure the consumer get a consistent result from the hook.
            dataRef.current = [(cachedValueRef.current ?? null) as CachedValue<TKey, TReturnData>, {status: newFetchStatus ?? 'loaded'}];
        }
```

> One question, why can't we use null value for the loading status as well? This way we could simplify the logic a bit here

I didn't understand quite well the question 😅 The comments on that condition was more directed to the use and difference between `cachedValueRef` and `dataRef`, the `newFetchStatus ?? 'loaded'` part I'm just defaulting to `loaded` in case it was not set before.

Reply:
Sorry! I meant why do we need to use undefined here:
```ts
        // If we have pending merge operations for the key during the first connection, we set the new value to `undefined`
        // and fetch status to `loading` to simulate that it is still being loaded until we have the most updated data.
        // If `allowStaleData` is `true` this logic will be ignored and cached value will be used, even if it's stale data.
        if (isFirstConnectionRef.current && Onyx.hasPendingMergeForKey(key) && !options?.allowStaleData) {
            newValue = undefined;
            newFetchStatus = 'loading';
        }
```

Most likely I'm missing something but if we were to use null in the logic above we could compare `dataRef` with `newData`, no?